### DISCUSSION TITLE
tutorials/2d/using_tilesets

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#220] [User-contributed notes]
# tutorials/2d/using_tilesets

Introduction: A tilemap is a grid of tiles used to create a game's layout. There are several benefits to using TileMapLayer nodes to design your levels. First, they let you draw a layout by "painti...

https://docs.godotengine.org/en/stable/tutorials/2d/using_tilesets.html

**Comment by boholstapp** [comment]
I use this with my TileMapLayer 

https://www.youtube.com/watch?v=7_fwNiZL0xY&t=59s

**Comment by nrichards** [comment]
The "missing docs". Explains using Terrain and peering bits:

https://github.com/dandeliondino/godot-4-tileset-terrains-docs

**Comment by space-jellyfish** [comment]
i think it'd be helpful if https://docs.godotengine.org/en/stable/classes/class_tilemap.html#class-tilemap-method-set-cell linked to here

**Comment by megonemad1** [comment]
Setting terrain from gd script was skipped over, and it's not really noted anywhere that you need to both set the cell and set_cells_terrain_connect to actually make terrain visible. The tilemaplayer doesn't handle that on your behalf.

**Comment by HybridDragon** [comment]
This Describes setting metadata for tile like a damage value. But actually getting that data in the context of a platformer seems convoluted.

**Comment by kolulu23** [comment]
_Texture Region Size_ is useful when you want to display a smaller texture in a tile,

**Comment by ozthekoder** [comment]
I cannot find a single example doing this with isometric tiles. Is creating terrains possible with isometric tiles?

**Comment by RedFox3DMaker** [comment]
Hello,
Is there a way to create more than one polygon for collision or navigation layers ? For instance, I have some tiles for a maze with walls so making the collision polygon is simple, but I couldn't find a way to create multiple disjoint polygon around it for the navigation. Is there another way to handle this case ?

**Reply by Brambleshadow11** [reply]
I was actually doing a map tileset myself! If you click "f" like in the tutorial, you can adjust that to fit on wall/corner. Then you need to click the green button with the +. That lets you add separate "nodes" unattached to the first. Just click on the first one once you've made your shape. And if you need to add more "nodes" to an exisitng completed shape, the blue button lets you add onto an existing shape.

**Comment by PangolinMontanari** [comment]
The use of the tile proxy system should be more clearly documented. The functionality is not self-explanatory.

**Comment by dginovker** [comment]
This page talks about assigning custom metadata for tiles well, but it doesn't mention how you use it in the code. How do the damage_per_second property of a tile a player stands on?

**Reply by Calinou** [reply]
You would need to check the the the player stands on by using [`TileMapLayer.local_to_map()`](https://docs.godotengine.org/en/latest/classes/class_tilemaplayer.html#class-tilemaplayer-method-local-to-map) (pass the player position as the parameter of that method). Do this in `_process()`to apply damage to the player whenever they are standing on a tile, or in a Timer's `timeout` signal function to apply it periodically.

**Reply by dginovker** [reply]
Thanks @Calinou  (and much love for the engine!). But TileMapLayer.local_to_map() implies the player is only standing on one tile. Is there a way I can get something like an on_enter signal so that I can apply mechanics based on custom metadata the moment the player touches the tile?

I.e. if my player is the Cyan square, its _position_ (the top left corner) isn't touching a tile in the below image; but I would still want to apply damage from the spike:

<img width="520" height="427" alt="image" src="https://github.com/user-attachments/assets/719709dd-9654-40b2-991b-5a396a775ad7" />

**Reply by Calinou** [reply]
There is no per-tile signal (that would be very slow), but you can use `local_to_map()` four times (once on each corner coordinate of your sprite). Then in your code, retain the *highest* damage per second value across all four corners and apply it to the player.

This should be fine performance-wise, but if performance were to become an issue because of this, you could always spread the checks to be performed across separate frames (e.g. check one corner per frame). You don't need to do this if it's just for the player and a few enemies, but if you have hundreds or thousands of enemies, it can start making sense.

**Comment by shaesung01-byte** [comment]
Why does the terrain just scroll to the right tile it's connected to? This game engine is killing me!

**Reply by LoganWyborn** [reply]
Can you elaborate a bit on your question or provide more info to help us understand what you mean? People are generally pretty open to helping out here, but we need enough information to actually understand your problem.

**Comment by Lumb3rJaX** [comment]
The section regarding the `Particles2s` scene creation is a bit confusing to me. Is this something that is strictly *necessary* in order to create the tileset? Or is this just something necessary for when effects are going to be used to display *over* a tileset.

**Comment by Funimac2014** [comment]
Three questions:
1. How can I make every block have collision without having to put a polygon collision hitbox over it?
2. How do I make ramp and ledge blocks that are triangular?
3. How do I make sections that I want to be see-through transparent?
(I use Microsoft Paint btw.)
<img width="672" height="747" alt="Ashbug_Nest_Terrain" src="https://github.com/user-attachments/assets/06b2e3c9-f011-4294-84e6-9714ab2ea75f" />

Extra question: How can I make hazards like spikes where if you touch them you take damage and get teleported to the nearest flat platform?

**Reply by rredesigns** [reply]
1 - What do you mean with "collision hitbox"? You can paint colliders on the tiles if you enable the physics layer on your TileSet, then just edit it to the shape you need.

2 - Just use the same tile size but use just a triangular collision shape for it that covers the ramp part.

3 - Your tiles themselves should be transparent. You should only be drawing details on a layer, not the background, unless those are base tiles (the ones at the bottom, everything else is drawn on top of them).

Extra: Just the same as the ramps, create your spikes tiles, give them appropriate collision shapes for the spikes and then script your damage/teleport event.
